spec/support/mock_twilio_client.rb

```diff
@@ -0,0 +1,37 @@
+class MockTwilioClient
+  Message = Struct.new(:from, :to, :body, :url)
+  Call = Struct.new(:from, :to, :body, :url)
+
+  cattr_accessor :messages
+  cattr_accessor :calls
+  self.messages = []
+  self.calls = []
+
+  def initialize(_account_sid, _auth_token)
+  end
+
+  def messages
+    @type = :messages
+    self
+  end
+
+  def calls
+    @type = :calls
+    self
+  end
+
+  # rubocop:disable all
+  # a metaprogramming method designed to capture #create calls
+  # for the Twilio::REST::Client and place the newly created struct
+  # in an array for verification
+  def create(from:, to:, body: nil, url: nil)
+    cmd = "#{@type.to_s.capitalize.singularize}.new(
+      from,
+      to,
+      body,
+      url)"
+    item = eval(cmd)
+    self.class.send(@type.to_s) << item
+  end
+  # rubocop:enable all
```

Comment:
In this case, I would use the dependency injection approach described in the thoughtbot article. We would modify the `TwilioService` class to add a class accessor method that allows us to specify which class to use:
```ruby
class TwilioService
  cattr_accessor :telephony_service do
    Twilio::REST::Client
  end

  ...

  def twilio_proxy_client
    telephony_service.new(
      account['sid'],
      account['auth_token'],
      proxy_addr: proxy_addr,
      proxy_port: proxy_port
    )
  end

  def twilio_client
    telephony_service.new(
      account['sid'],
      account['auth_token']
    )
  end
  ...
end
```
Then, in each test, you can specify the approprate fake class:
```ruby
# voice_sender_otp_job_spec.rb

describe VoiceSenderOtpJob do
  describe '.perform' do
    it 'initiates the phone call to deliver the OTP', twilio: true do
      TwilioService.telephony_service = FakeVoiceCall

      VoiceSenderOtpJob.perform_now('1234', '555-5555')

      calls = FakeVoiceCall.calls

      code = '1234'.scan(/\d/).join(', ')
      message = t('voice.otp_confirmation', code: code)
      url_message = URI.escape(message)

      expect(calls.size).to eq(1)
      call = calls.first
      expect(call.to).to eq('555-5555')
      expect(call.url).to include(url_message)
    end
  end
end
```
There would be a separate fake class for each type of telephony service:
```ruby
class FakeVoiceCall
  Call = Struct.new(:from, :to, :url)

  cattr_accessor :calls
  self.calls = []

  def initialize(_account_sid, _auth_token)
  end

  def calls
    self
  end

  def create(params = {})
    self.class.calls << Call.new(params[:from], params[:to], params[:url])
  end
end
```
```ruby
class FakeSms
  Message = Struct.new(:from, :to, :body)

  cattr_accessor :messages
  self.messages = []

  def initialize(_account_sid, _auth_token)
  end

  def messages
    self
  end

  def create(params = {})
    self.class.messages << Message.new(params[:from], params[:to], params[:body])
  end
end
```
And in the `rails_helper`, you would initialize the arrays for each class:
```ruby
config.before(:each, twilio: true) do
  FakeSms.messages = []
  FakeVoiceCall.calls = []
end
```
Note that I chose the name `Fake` instead of `Mock` for consistency with other fake classes we already have.